src/java/org/apache/cassandra/transport/messages/StartupMessage.java

```diff
@@ -118,8 +121,34 @@ else if (compression.equals("lz4"))
             clientState.setDriverVersion(options.get(DRIVER_VERSION));
         }
 
-        if (DatabaseDescriptor.getAuthenticator().requireAuthentication())
+        IAuthenticator authenticator = DatabaseDescriptor.getAuthenticator();
+        if (authenticator.requireAuthentication())
+        {
+            // If the authenticator supports early certificate authentication, attempt to authenticate with certificates.
+            if (authenticator.supportsEarlyCertificateAuthentication())
```

Comment:
in the mtls with password auth fallback, we could also authenticate here if client certs are presented . I'm thinking the API could be a bit more generic, thinking of supporting other types of early authentication .

```
if (authenticator.supportsEarlyAuthentication(someContextThatHelpsMakingTheDetermination))
```

Reply:
If I'm understanding correctly, I think that is the way it should work with the current state of things:

`MutualTlsWithPasswordFallbackAuthenticator.supportsEarlyCertificateAuthentication()` will always return `true` 

The `SaslNegotiator` returned from `MutualTlsWithPasswordFallbackAuthenticator.newSaslNegotiator` will return an implementation that evaluates to `false` for `requiresCertificateAuthentication()` if no certs are present on the connection, and `true` if certificates are present which will then cause it to attempt to authenticate (line 135 below).

--

I do like the idea of more generically naming this though;  Are there other kinds of authentication that you could envision of that we could do eventually without sending an AUTHENTICATE message?  Other things present on the connection that we could validate with?

I think if we renamed:

1. `IAuthenticator.supportsEarlyCertificateAuthentication` to `IAuthenticator.supportsEarlyAuthentication`
2. `SaslNegotiator.requiresCertificateAuthentication` to `SaslNegotiation.requiresEarlyAuthentication`

That could work?  In `AuthUtil.handleLogin` we pass `connection` and `state`, which should be the level of context needed to do other authentication (unless there is something else we can pass through that would be helpful for authentication)

Reply:
> The SaslNegotiator returned from MutualTlsWithPasswordFallbackAuthenticator.newSaslNegotiator will return an implementation that evaluates to false for requiresCertificateAuthentication() if no certs are present on the connection, and true if certificates are present which will then cause it to attempt to authenticate (line 135 below).

Ah yes, I see that now, it is based on whether the certificates are present. This makes sense

> Are there other kinds of authentication that you could envision of that we could do eventually without sending an AUTHENTICATE message?

I was trying to think of other use cases, but I couldn't come up with anything for now. But I think making the method name generic could open it up for future improvements (maybe). We don't really need to _tie_ the implementation to certificates.

+1 to renaming it as proposed above

Reply:
agreed!  I'll rename the methods and push a commit imminently

Reply:
Pushed a change for this [f115a89](https://github.com/apache/cassandra/pull/2969/commits/f115a89605673463ccfaac600057fceb13917f1b)

I made one slight change to my proposal, instead of 

> SaslNegotiator.requiresCertificateAuthentication to SaslNegotiation.requiresEarlyAuthentication

I removed the method and created `SaslNegotiator.shouldSendAuthenticateMessage()` which indicates whether or not a further AUTHENTICATE message should be sent to the client, I felt that was more clear than my earlier proposal.  I am open to changing that though.  Hopefully this looks ok!